### ISSUE TITLE
Union[NoReturn, A] doesn't get narrowed to A

### ISSUE THREAD

**Issue by decorator-factory** [microsoft/pyright#1308]
```py
class Foo:
    def __iter__(self):
        return self

    def __next__(self):
        raise ValueError("No foos available :-(")

def f(foo: Foo):
    a, b = foo
    print(len(a))
# Argument of type "NoReturn" cannot be assigned to parameter "__obj" of type "Sized" in function "len"
#    "__len__" is not present


class Bar:
    def __iter__(self):
        return self

    def __next__(self):
        return "bar"

def g(foobar: Union[Foo, Bar]):
    a, b = foobar
    print(len(a))
# Argument of type "NoReturn | Literal['bar']" cannot be assigned to parameter "__obj" of type "Sized" in function "len"
#  Type "NoReturn | Literal['bar']" cannot be assigned to type "Sized"
#    "__len__" is not present
```
The first one (`f`) is kind of understandable, although `print(len(a))` should rather be marked as unreachable, right?
The second one (`g`) seems wrong. If we've reached the `print(len(a))` line, then clearly `a` is `Literal["bar"]`, not `NoReturn`

**Comment by erictraut** [comment]
There are arguably bugs in both of these samples. You are iterating over the instance of a type that does not support iteration and instead raises exceptions. If you fix the bugs in your code, these errors will go away. If you don't want to fix your bug, you can work around the issue by adding an explicit return type annotation to the `__next__` method.

**Comment by erictraut** [comment]
I've fixed the issue with the NoReturn appearing in a union resulting from the evaluation of an iterable. This will be in the next release.

**Comment by decorator-factory** [comment]
> There are arguably bugs in both of these samples. You are iterating over the instance of a type that does not support iteration and instead raises exceptions. If you fix the bugs in your code, these errors will go away. If you don't want to fix your bug, you can work around the issue by adding an explicit return type annotation to the `__next__` method.

Yeah, these examples are pretty silly. This is what led me to find this:
I was implementing a stack-based language, and I store stacks as `Stack = Optional[tuple[Value, Stack]]`.
In built-in functions, I unpacked stacks like this:
```py
(z, (y, (x, rest))) = stack
```
if unpacking goes wrong, I get a nasty error at runtime that I can't really catch. So I wanted to make a `StackBottom` type that throws a special `StackUnderflow` exception on `__iter__` or `__next__`. I suppose I can fix that by assigning `-> Value` to the `__next__`.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.97, which I just published. It will also be included in the next release of pylance.
